src/docfx/restore/index/DependencyGitPool.cs

```diff
@@ -0,0 +1,175 @@
+        {
+            Debug.Assert(dependencyLock != null);
+
+            var gitVersion = dependencyLock.GetGitLock(remote, branch);
+
+            if (gitVersion == null)
+            {
+                throw Errors.NeedRestore($"{remote}#{branch}").ToException();
+            }
+
+            if (!_acquiredGits.TryGetValue((remote, branch, gitVersion.Commit), out var gitInfo))
+            {
+                throw Errors.NeedRestore($"{remote}#{branch}").ToException();
+            }
+
+            if (string.IsNullOrEmpty(gitInfo.path) || gitInfo.git == null)
+            {
+                throw Errors.NeedRestore($"{remote}#{branch}").ToException();
+            }
+
+            var path = Path.Combine(AppData.GetGitDir(remote), gitInfo.path);
+            Debug.Assert(Directory.Exists(path));
+
+            return (path, gitVersion);
+        }
+
+        public async Task<bool> Release()
+        {
+            var released = true;
+            foreach (var (k, v) in _acquiredGits)
+            {
+                released &= await DependencySlotPool.ReleaseSlot(v.git, LockType.Shared);
+            }
+
+            Debug.Assert(released);
+
+            return released;
+        }
+
+        /// <summary>
+        /// Acquired all shared git based on dependency lock
+        /// The dependency lock must be loaded before using this method
+        /// </summary>
+        public static async Task<DependencyGitPool>
+            AcquireGitPool(
+            DependencyLockModel dependencyLock,
+            Dictionary<(string remote, string branch, string commit), (string path, DependencyGit git)> acquired = null)
+        {
+            Debug.Assert(dependencyLock != null);
+
+            var root = acquired == null;
+            acquired = acquired ?? new Dictionary<(string remote, string branch, string commit), (string path, DependencyGit git)>();
+
+            var successed = true;
+            try
+            {
+                foreach (var gitVersion in dependencyLock.Git)
+                {
+                    var (remote, branch, _) = HrefUtility.SplitGitHref(gitVersion.Key);
+                    if (!acquired.ContainsKey((remote, branch, gitVersion.Value.Commit/*commit*/)))
+                    {
+                        var (path, git) = await AcquireGit(remote, branch, gitVersion.Value.Commit, LockType.Shared);
+                        acquired[(remote, branch, gitVersion.Value.Commit/*commit*/)] = (path, git);
+                    }
+
+                    await AcquireGitPool(gitVersion.Value, acquired);
+                }
+            }
+            catch
+            {
+                successed = false;
+                throw;
+            }
+            finally
+            {
+                if (!successed && root)
+                {
+                    foreach (var (k, v) in acquired)
```

Comment:
Can we just call `Release` here?
Since `AcquireGitPool` is supposed to be invoked only once to get the instance?

Reply:
offline discussed